### ISSUE TITLE
Feature Request - Add ROS package.xml

### ISSUE THREAD

**Issue by Ryanf55** [Exiv2/exiv2#2724]
#### **Is your feature request related to a problem?**

When consuming exiv2 in [ROS 2](https://docs.ros.org/en/rolling/index.html), it's standard practice to add a `package.xml` next to the top level CMakeLists of the project. Without it, dependencies must be manually installed, and build order manually maintained. Since ROS 2 does not yet support `conan`, `rosdep` can be viewed as an alternative. 

#### **Describe the solution you would like**

* Add a `package.xml` to the root of this repo that details build and test dependencies for the default configuration options.
* Submit upstream [rosdep](https://docs.ros.org/en/rolling/Tutorials/Intermediate/Rosdep.html) PR (s) for any dependencies that do not already exist.

#### **Describe alternatives you have considered**

* Manage exiv2's dependencies manually in ROS 2
* Manage the build order of exiv2 relative to its consuming libraries manually in ROS 2
* Use an alternative package manager like `conan` - not supported by `colcon` yet

#### **Desktop**

- OS and version: Ubuntu 22.04, and others supported by rosdep
- Exiv2 version and source: 0.28.x
- Any software using exiv2 and source: closed source, but available for use with `rosdep` today due to some previous work: https://github.com/ros/rosdistro/pull/37226

#### **Additional context**

* When installing dependencies with a `ros` based system, the tool `rosdep` is used, which reads the `package.xml` manifest and installs dependencies. It is OS-portable between many Linux distros, Mac, and Windows.
* When building multiple packages with the `colcon` build tool, the presence of a `package.xml` informs `colcon` on which order to build the packages
* A working CMake installation that supports `find_package` is a pre-requisite to using exiv2 in a `colcon` build system. Since that is working, there are no blockers to doing this work.
* Adding a `package.xml` should not break anything. It's a standalone file, so I'd request once this is merged, it is backported to 0.28.x. 
* I can do the work for this PR, I would just like approval of the idea and it's something the `exiv2` maintainers would support merging before I start the work.

**Comment by Ryanf55** [comment]
> I don't think this is needed. Exiv2 is normally available in Ubuntu distro packages. colcon can normally use any properly installed system packages. What you might have been missing is the relevant -dev package adding the development libraries. That will be added shortly by [ros/rosdistro#43463](https://github.com/ros/rosdistro/pull/43463) . You then just `<build_depend>exiv2-dev</build_depend>` and `find_package(exiv2 REQUIRED)` and that's it, you can use Exiv2 API in your code.

That's assuming that the version released by the ubuntu maintainers is bug free and has all the features needed by the application. It doesn't. 0.28.x is a large improvement over 0.27.x. The fundamental difference between the ROS build farm and  ubuntu distributed packages is the release theory. ROS allows ABI-preserving changes in regularly to stable branches (humble) and releases regularly. Ubuntu only allows critical bugfixes, and, from my experience, the entire ubuntu distribution is severely understaffed and undermaintained, resulting in many known bugs on Ubuntu 22 caused by the distribution that are fixed in newer, ABI compatible versions. 

The use of a package.xml allows people like me to pick which version of exiv2 to use (likely newer) and not be held back bu the Ubuntu packaging process.
> 
> > * Any software using exiv2 and source: closed source,
> 
> Uh-oh, Exiv2 is GPL-licensed. You can't build closed-source programs on top of it when using its API. The only way (but it's a legal grey zone) is using the CLI tools by spawning a subprocess and parsing its output.

Sure you can! You just have to license available to the customer, and a written offer to make the source code, including anything that links to the GPL code, available to your customer. Plus, if it's not used in a product, such as only used in an internal R&D tool that's never distributed, then there is nothing to worry about. The nice thing about adding a package.xml is that it more quickly allows a software team or product manager to issue a report on which licenses (software bill of materials) are used by which package, which packages depend on work that is GPL2, and make sure that it never reaches a product that we aren't comfortable releasing all the code for. There was a nice ROSCon talk from last year about the new licensing changes.
https://roscon.ros.org/2023/talks/Licensing_of_ROS_Packages_Changes_to_package.xml_and_new_tool_support.pdf

**Comment by peci1** [comment]
> That's assuming that the version released by the ubuntu maintainers is bug free and has all the features needed by the application.

I share your pain about this, but when you release a ROS package on buildfarm, it has to work with the stock versions of the libraries. That's the price of using a distribution. On the other hand, the distribution is a more or less cohesive set of packages that should be able to interoperate. If you need a different version, you need to build it inside your package and statically link it.

There were very few exceptions to this like FCL, which got built in a different version by ROS buildfarm, so you could install both libfcl-dev and ros-noetic-fcl at one system. But the number of headaches this brought showed to be quite large.

> The use of a package.xml allows people like me to pick which version of exiv2 to use (likely newer) and not be held back by the Ubuntu packaging process.

You are free to do this even without package.xml in exiv2. Just build the library and install it into a path where your workspace CMake will find it.

I know it would be easier if package.xml were in exiv2, because then you'd just drop it into your workspace, run rosdep and build and you'd be done. However, it doesn't seem like a good path to me. Would you want to add package.xml also to GCC, GLibc, Python, libusb or any other library a ROS user might potentially want to use in a non-default version? The risk of doing this is that the authors of these "unrelated" libraries know nothing about ROS and rosdep, so they will most likely do changes in the future that will not be reflected in the package.xml. Then you'd end up with a broken system.

If you absolutely need an automated way of getting the library built by your workspace, you might consider distributing a custom rosdep rule. They can use bash scripts to install/check presence of packages.

> ROS allows ABI-preserving changes in regularly to stable branches (humble) and releases regularly.

Just a side note: Only Quality-Level 1 packages have this limitation. There's a bunch of packages that break ABI regularly.

> Sure you can! You just have to license available to the customer, and a written offer to make the source code, including anything that links to the GPL code, available to your customer.

Okay, that's a point of view I never imagined, but probably valid. But you can't restrict your customers to publicize the GPL'd parts, right?

> The nice thing about adding a package.xml is that it more quickly allows a software team or product manager to issue a report on which licenses (software bill of materials) are used by which package

This sounds like you're aiming at the wrong tool. package.xml is a nice interface for viewing and collecting licenses, but it will never cover your whole software stack. You should be looking into Debian copyright declarations. Each released package has to have it (and you can only hope it is correct, which is, however, the same with package.xml). Or, in some nice future, you could probably also use a REUSE-linter to check for SPDX licences of all included files.

---

To sum it up, I don't think there is any good reason to push package.xml files into non-ROS-releated projects.

**Comment by Ryanf55** [comment]
Great points, I appreciate the discussion!



> I share your pain about this, but when you release a ROS package on buildfarm, it has to work with the stock versions of the libraries. 

We don't release our internal dev tools that rely on exiv2 to the build farm so I don't share that motivation in this case.


> 
> You are free to do this even without package.xml in exiv2. Just build the library and install it into a path where your workspace CMake will find it.

This requires separate build infrastructure, which setting up is of little customer value. An agglomeration of shell scripts and dpkg-deb and custom invocations of CMake CLI along with a place to host the debians is the best we have, and the only reason we do that is to save time on clean builds where we exlude dependencies like exiv2 that hardly change.

The only reason we don't put it in the workspace now is because `colcon` doesn't know to build it first becuase it's missing a package.xml. Why reinvent the build system for some dependencies. I don't see it much different than supporting `conan` recipes or `yocto` layers, assuming these `package.xml` that are traditionally community maintained. The ROS community just doesn't have a place to put these package.xml's because they have to live in the repo.

> I know it would be easier if package.xml were in exiv2, because then you'd just drop it into your workspace, run rosdep and build and you'd be done. However, it doesn't seem like a good path to me. Would you want to add package.xml also to GCC, GLibc, Python, libusb or any other library a ROS user might potentially want to use in a non-default version? The risk of doing this is that the authors of these "unrelated" libraries know nothing about ROS and rosdep, so they will most likely do changes in the future that will not be reflected in the package.xml. Then you'd end up with a broken system.

Ditto above, I think this is really a limitation of rosdep/colcon requiring the `package.xml` to be in the same directory as the source tree.

> If you absolutely need an automated way of getting the library built by your workspace, you might consider distributing a custom rosdep rule. They can use bash scripts to install/check presence of packages.

Got any examples/details?  We don't need it, I'm just trying to avoid reinventing the build system for our non-ROS packages.


> Okay, that's a point of view I never imagined, but probably valid. But you can't restrict your customers to publicize the GPL'd parts, right?

Nope, we can't. I'm not an expert on this or a lawyer, so I'm not willing to provide public recommendations. 

> > The nice thing about adding a package.xml is that it more quickly allows a software team or product manager to issue a report on which licenses (software bill of materials) are used by which package
> 
> This sounds like you're aiming at the wrong tool. package.xml is a nice interface for viewing and collecting licenses, but it will never cover your whole software stack. You should be looking into Debian copyright declarations. Each released package has to have it (and you can only hope it is correct, which is, however, the same with package.xml). Or, in some nice future, you could probably also use a REUSE-linter to check for SPDX licences of all included files.
> 
> To sum it up, I don't think there is any good reason to push package.xml files into non-ROS-releated projects.

I appreciate the points, and think you've made a valid enough case (and judging by the lack of engagement from the maintainers at a community-maintained package.xml in the repo), I'll close this. I'd be happy to follow up on ROS discourse. I think there's improvements to make in the documentation for 
1) How to use newer non-ROS dependencies in a ROS ecosystem for private use where you don't have to worry about the open source build farm
2) How to collect licenses in a ROS ecosystem that uses a mix of ROS packages and system dependencies

**Comment by peci1** [comment]
Thanks for the great discussion. I'd be happy to continue on ROS discourse. Many of your points are quite relevant. Having the possibility to provide external package.xml would help in some cases.

> > If you absolutely need an automated way of getting the library built by your workspace, you might consider distributing a custom rosdep rule. They can use bash scripts to install/check presence of packages.
> 
> Got any examples/details? We don't need it, I'm just trying to avoid reinventing the build system for our non-ROS packages.

Put your file in `/etc/ros/rosdep/sources.list.d` . An example of the files to put there is e.g. https://github.com/basler/pylon-ros-camera/tree/master/pylon_camera/rosdep .

**Comment by Ryanf55** [comment]
Follow up: https://discourse.ros.org/t/leveraging-ros-build-infrastructure-for-non-ros-packages/40658
